File: Day-4-linked-list/javascript/singly_linked_list.js
class Node {
	constructor(data) {
	  this.data = data;
	  this.next = null;
	}
  }
  
  class LinkedList {
	constructor() {
	  this.head = null;
	}
  
	insertAtBeginning(data) {
	  const newNode = new Node(data);
	  newNode.next = this.head;
	  this.head = newNode;
	}
  
	insertAtEnd(data) {
	  const newNode = new Node(data);
	  if (!this.head) {
		this.head = newNode;
		return;
	  }
	  let last = this.head;
	  while (last.next) {
		last = last.next;
	  }
	  last.next = newNode;
	}
  
	insertAtPosition(position, data) {
	  if (position < 0) {
		console.log("Position can't be negative.");
		return;
	  }
	  const newNode = new Node(data);
	  if (position === 0) {
		newNode.next = this.head;
		this.head = newNode;
		return;
	  }
	  let temp = this.head;
	  for (let i = 0; i < position - 1; i++) {
		if (!temp) {
		  console.log("Position is greater than the length of the list.");
		  return;
		}
		temp = temp.next;
	  }
	  if (!temp) {
		console.log("Position is greater than the length of the list.");
		return;
	  }
	  newNode.next = temp.next;
	  temp.next = newNode;
	}
  
	deleteNode(key) {
	  let temp = this.head;
	  if (temp && temp.data === key) {
		this.head = temp.next;
		temp = null;
		return;
	  }
	  let prev = null;
	  while (temp && temp.data !== key) {
		prev = temp;
		temp = temp.next;
	  }
	  if (!temp) {
		return;
	  }
	  prev.next = temp.next;
	  temp = null;
	}
  
	printList() {
	  let temp = this.head;
	  while (temp) {
		process.stdout.write(temp.data + ' ');
		temp = temp.next;
	  }
	  console.log();
	}
  }
  
  // Usage
  const llist = new LinkedList();
  llist.insertAtEnd(1);
  llist.insertAtEnd(2);
  llist.insertAtBeginning(0);
  llist.printList();  // Output: 0 1 2
  llist.deleteNode(1);
  llist.printList();  // Output: 0 2
  llist.insertAtPosition(1, 1);
  llist.printList();  // Output: 0 1 2
  

File: Day-4-linked-list/golang/singly_linked_list.go
package main

import "fmt"

type Node struct {
	data int
	next *Node
}

type LinkedList struct {
	head *Node
}

func (ll *LinkedList) InsertAtBeginning(data int) {
	newNode := &Node{data: data}
	newNode.next = ll.head
	ll.head = newNode
}

func (ll *LinkedList) InsertAtEnd(data int) {
	newNode := &Node{data: data}
	if ll.head == nil {
		ll.head = newNode
		return
	}
	last := ll.head
	for last.next != nil {
		last = last.next
	}
	last.next = newNode
}

func (ll *LinkedList) InsertAtPosition(position int, data int) {
	if position < 0 {
		fmt.Println("Position can't be negative.")
		return
	}
	newNode := &Node{data: data}
	if position == 0 {
		newNode.next = ll.head
		ll.head = newNode
		return
	}
	temp := ll.head
	for i := 0; i < position-1; i++ {
		if temp == nil {
			fmt.Println("Position is greater than the length of the list.")
			return
		}
		temp = temp.next
	}
	if temp == nil {
		fmt.Println("Position is greater than the length of the list.")
		return
	}
	newNode.next = temp.next
	temp.next = newNode
}

func (ll *LinkedList) DeleteNode(key int) {
	temp := ll.head
	if temp != nil && temp.data == key {
		ll.head = temp.next
		temp = nil
		return
	}
	var prev *Node
	for temp != nil && temp.data != key {
		prev = temp
		temp = temp.next
	}
	if temp == nil {
		return
	}
	prev.next = temp.next
	temp = nil
}

func (ll *LinkedList) PrintList() {
	temp := ll.head
	for temp != nil {
		fmt.Print(temp.data, " ")
		temp = temp.next
	}
	fmt.Println()
}

func main() {
	llist := &LinkedList{}
	llist.InsertAtEnd(1)
	llist.InsertAtEnd(2)
	llist.InsertAtBeginning(0)
	llist.PrintList() // Output: 0 1 2
	llist.DeleteNode(1)
	llist.PrintList() // Output: 0 2
	llist.InsertAtPosition(1, 1)
	llist.PrintList() // Output: 0 1 2
}
